File: sodp-go/frame.go
// Package sodp implements the State-Oriented Data Protocol (SODP) v1.0
// as an embeddable Go library for real-time state synchronization over WebSockets.
//
// Wire format: each message is a 4-element MessagePack array:
//
//	[frame_type(u8), stream_id(u32), seq(u64), body(any)]
//
// Quickstart:
//
//	srv := sodp.NewServer()
//	http.HandleFunc("/sodp", srv.HandleWS)
//	http.ListenAndServe(":7777", nil)
package sodp

import (
	"fmt"

	"github.com/vmihailenco/msgpack/v5"
)

// FrameType identifies the SODP message type (wire: u8).
type FrameType uint8

const (
	FrameHello     FrameType = 0x01
	FrameWatch     FrameType = 0x02
	FrameStateInit FrameType = 0x03
	FrameDelta     FrameType = 0x04
	FrameCall      FrameType = 0x05
	FrameResult    FrameType = 0x06
	FrameError     FrameType = 0x07
	FrameHeartbeat FrameType = 0x09
	FrameResume    FrameType = 0x0A
	FrameAuth      FrameType = 0x0B
	FrameAuthOK    FrameType = 0x0C
	FrameUnwatch   FrameType = 0x0D
)

// Frame is a decoded SODP wire message.
type Frame struct {
	Type     FrameType
	StreamID uint32
	Seq      uint64
	Body     any
}

// EncodeFrame serializes a Frame into MessagePack bytes.
func EncodeFrame(f Frame) ([]byte, error) {
	return msgpack.Marshal([]any{uint8(f.Type), f.StreamID, f.Seq, f.Body})
}

// DecodeFrame deserializes MessagePack bytes into a Frame.
func DecodeFrame(data []byte) (Frame, error) {
	var raw []msgpack.RawMessage
	if err := msgpack.Unmarshal(data, &raw); err != nil {
		return Frame{}, fmt.Errorf("sodp: invalid frame: %w", err)
	}
	if len(raw) != 4 {
		return Frame{}, fmt.Errorf("sodp: frame must have 4 elements, got %d", len(raw))
	}

	var ft uint8
	if err := msgpack.Unmarshal(raw[0], &ft); err != nil {
		return Frame{}, fmt.Errorf("sodp: invalid frame_type: %w", err)
	}
	var streamID uint32
	if err := msgpack.Unmarshal(raw[1], &streamID); err != nil {
		return Frame{}, fmt.Errorf("sodp: invalid stream_id: %w", err)
	}
	var seq uint64
	if err := msgpack.Unmarshal(raw[2], &seq); err != nil {
		return Frame{}, fmt.Errorf("sodp: invalid seq: %w", err)
	}

	var body any
	if len(raw[3]) > 0 {
		if err := msgpack.Unmarshal(raw[3], &body); err != nil {
			return Frame{}, fmt.Errorf("sodp: invalid body: %w", err)
		}
	}

	return Frame{
		Type:     FrameType(ft),
		StreamID: streamID,
		Seq:      seq,
		Body:     body,
	}, nil
}

// HelloBody is the server's initial handshake payload.
type HelloBody struct {
	Protocol     string         `msgpack:"protocol"`
	Version      string         `msgpack:"version"`
	ServerID     string         `msgpack:"server_id"`
	Auth         bool           `msgpack:"auth"`
	Capabilities map[string]any `msgpack:"capabilities,omitempty"`
}

// WatchBody is the client's subscription request (WATCH and RESUME).
type WatchBody struct {
	Key          string         `msgpack:"state"`
	States       []string       `msgpack:"states,omitempty"`
	SinceVersion uint64         `msgpack:"since_version,omitempty"`
	Params       map[string]any `msgpack:"params,omitempty"`
}

// CallBody is the CALL frame payload.
type CallBody struct {
	CallID string         `msgpack:"call_id"`
	Method string         `msgpack:"method"`
	Args   map[string]any `msgpack:"args"`
}

// ErrorBody carries structured error details.
type ErrorBody struct {
	Code    int    `msgpack:"code"`
	Message string `msgpack:"message"`
}

// AuthBody carries the JWT token from the client.
type AuthBody struct {
	Token string `msgpack:"token"`
}

// AuthOKBody confirms authentication success.
type AuthOKBody struct {
	Subject string `msgpack:"sub"`
}

File: server-ts/src/frame.ts
// SODP v1.0 wire framing.
//
// Every frame is a 4-tuple MessagePack array: [op, stream, seq, body].
// Clients use @sodp/client which talks this exact shape.
//
// Opcode ranges:
//   0x01–0x0D control/data frames (HELLO..UNWATCH)

import { encode, decode } from '@msgpack/msgpack'

export const OP = {
  HELLO:      0x01,
  WATCH:      0x02,
  STATE_INIT: 0x03,
  DELTA:      0x04,
  CALL:       0x05,
  RESULT:     0x06,
  ERROR:      0x07,
  HEARTBEAT:  0x09,
  RESUME:     0x0A,
  AUTH:       0x0B,
  AUTH_OK:    0x0C,
  UNWATCH:    0x0D,
} as const

export type Op = typeof OP[keyof typeof OP]

export interface Frame<TBody = unknown> {
  op: Op
  stream: number
  seq: number
  body: TBody
}

export interface HelloBody {
  protocol: 'sodp'
  version: '1.0'
  server: string
  auth: boolean
  capabilities: {
    rate_limit_writes: number
    rate_limit_watches: number
    backpressure_limit: number
    multi_watch: boolean
    params: boolean
    resume: boolean
  }
}

export interface StateInitBody {
  state: string
  version: number
  value: unknown
  initialized: boolean
  params?: Record<string, unknown>
}

export interface DeltaBody {
  version: number
  ops: DeltaOp[]
  state?: string
}

export interface DeltaOp {
  op: 'ADD' | 'UPDATE' | 'REMOVE'
  path: string
  value?: unknown
}

export interface ErrorBody {
  code: number
  message: string
}

export interface AuthOkBody {
  sub: string
}

export interface WatchBody {
  state?: string
  states?: string[]
  params?: Record<string, unknown>
}

export interface UnwatchBody {
  state?: string
  states?: string[]
}

export interface ResumeBody {
  state: string
  since_version: number
  params?: Record<string, unknown>
}

export interface AuthBody {
  token: string
}

export interface CallBody {
  call_id: string
  method: string
  args: Record<string, unknown>
}

function encodeFrame(op: Op, stream: number, seq: number, body: unknown): Uint8Array {
  return encode([op, stream, seq, body])
}

export const Frames = {
  hello(body: HelloBody): Uint8Array {
    return encodeFrame(OP.HELLO, 0, 0, body)
  },
  authOk(seq: number, body: AuthOkBody): Uint8Array {
    return encodeFrame(OP.AUTH_OK, 0, seq, body)
  },
  stateInit(stream: number, seq: number, body: StateInitBody): Uint8Array {
    return encodeFrame(OP.STATE_INIT, stream, seq, body)
  },
  delta(stream: number, seq: number, body: DeltaBody): Uint8Array {
    return encodeFrame(OP.DELTA, stream, seq, body)
  },
  error(stream: number, seq: number, body: ErrorBody): Uint8Array {
    return encodeFrame(OP.ERROR, stream, seq, body)
  },
  heartbeat(): Uint8Array {
    return encodeFrame(OP.HEARTBEAT, 0, 0, null)
  },
  result(seq: number, callId: string, success: boolean, data: unknown): Uint8Array {
    return encodeFrame(OP.RESULT, 0, seq, { call_id: callId, success, data })
  },
} as const

export function decodeFrame(buf: ArrayBufferLike | Uint8Array): Frame {
  const raw = decode(buf)
  if (!Array.isArray(raw) || raw.length !== 4) {
    throw new Error('malformed frame: not a 4-tuple')
  }
  const [op, stream, seq, body] = raw as [unknown, unknown, unknown, unknown]
  if (typeof op !== 'number' || typeof stream !== 'number' || typeof seq !== 'number') {
    throw new Error('malformed frame: op/stream/seq must be numbers')
  }
  return { op: op as Op, stream, seq, body }
}
